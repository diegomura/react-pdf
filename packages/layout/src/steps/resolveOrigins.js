import getOrigin from '../node/getOrigin';

/**
 * Resolve node origin
 *
 * @param {Object} node
 * @returns {Object} node with origin attribute
 */
const resolveNodeOrigin = node => {
  const origin = getOrigin(node);

  const newNode = Object.assign({}, node, { origin });

  if (!node.children) return newNode;

  const children = node.children.map(resolveNodeOrigin);

  return Object.assign({}, newNode, { children });
};

/**
 * Resolve document origins
 *
 * @param {Object} document root
 * @returns {Object} documrnt root
 */

const resolveOrigin = root => {
  if (!root.children) return root;

  const children = root.children.map(resolveNodeOrigin);

  return Object.assign({}, root, { children });
};

export default resolveOrigin;
