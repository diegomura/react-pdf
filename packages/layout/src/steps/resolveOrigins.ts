import getOrigin from '../node/getOrigin';
import { SafeDocumentNode, SafeNode } from '../types';

/**
 * Resolve node origin
 *
 * @param node
 * @returns Node with origin attribute
 */
const resolveNodeOrigin = (node: SafeNode): SafeNode => {
  const origin = getOrigin(node);

  const newNode = Object.assign({}, node, { origin });

  if (!node.children) return newNode;

  const children = node.children.map(resolveNodeOrigin);

  return Object.assign({}, newNode, { children });
};

/**
 * Resolve document origins
 *
 * @param root - Document root
 * @returns Document root
 */

const resolveOrigin = (root: SafeDocumentNode) => {
  if (!root.children) return root;

  const children = root.children.map(resolveNodeOrigin);

  return Object.assign({}, root, { children });
};

export default resolveOrigin;
