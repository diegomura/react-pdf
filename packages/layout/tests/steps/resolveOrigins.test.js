import resolveOrigins from '../../src/steps/resolveOrigins';

describe('layout resolveOrigins', () => {
  test('should not resolve for node without box', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          style: {},
          children: [{ type: 'VIEW', style: {} }],
        },
      ],
    };
    const result = resolveOrigins(root);

    expect(result).toMatchSnapshot();
  });

  test('should resolve centered origin by default', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          style: {},
          box: { top: 50, left: 100, width: 300, height: 400 },
        },
      ],
    };
    const result = resolveOrigins(root);

    expect(result).toMatchSnapshot();
  });

  test('should resolve origin adjusted by fixed values', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          box: { top: 50, left: 100, width: 300, height: 400 },
          style: { transformOriginX: 100, transformOriginY: 50 },
        },
      ],
    };
    const result = resolveOrigins(root);

    expect(result).toMatchSnapshot();
  });

  test('should resolve origin adjusted by percent values', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          box: { top: 50, left: 100, width: 300, height: 400 },
          style: { transformOriginX: '20%', transformOriginY: '70%' },
        },
      ],
    };
    const result = resolveOrigins(root);

    expect(result).toMatchSnapshot();
  });

  test('should resolve origins for nested elements', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          style: {},
          box: { top: 50, left: 100, width: 300, height: 400 },
          children: [
            {
              type: 'VIEW',
              style: {},
              box: { top: 0, left: 10, width: 50, height: 80 },
            },
          ],
        },
      ],
    };
    const result = resolveOrigins(root);

    expect(result).toMatchSnapshot();
  });
});
