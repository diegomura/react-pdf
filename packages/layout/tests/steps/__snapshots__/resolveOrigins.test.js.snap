// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout resolveOrigins should not resolve for node without box 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "origin": Object {},
          "style": Object {},
          "type": "VIEW",
        },
      ],
      "origin": Object {},
      "style": Object {},
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveOrigins should resolve centered origin by default 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 400,
        "left": 100,
        "top": 50,
        "width": 300,
      },
      "origin": Object {
        "left": 250,
        "top": 250,
      },
      "style": Object {},
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveOrigins should resolve origin adjusted by fixed values 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 400,
        "left": 100,
        "top": 50,
        "width": 300,
      },
      "origin": Object {
        "left": 200,
        "top": 100,
      },
      "style": Object {
        "transformOriginX": 100,
        "transformOriginY": 50,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveOrigins should resolve origin adjusted by percent values 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 400,
        "left": 100,
        "top": 50,
        "width": 300,
      },
      "origin": Object {
        "left": 160,
        "top": 330,
      },
      "style": Object {
        "transformOriginX": "20%",
        "transformOriginY": "70%",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveOrigins should resolve origins for nested elements 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 400,
        "left": 100,
        "top": 50,
        "width": 300,
      },
      "children": Array [
        Object {
          "box": Object {
            "height": 80,
            "left": 10,
            "top": 0,
            "width": 50,
          },
          "origin": Object {
            "left": 35,
            "top": 40,
          },
          "style": Object {},
          "type": "VIEW",
        },
      ],
      "origin": Object {
        "left": 250,
        "top": 250,
      },
      "style": Object {},
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;
